File: 323__Number_of_Connected_Components_in_an_Undirected_Graph.py
class Solution(object):
    def countComponents(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: int
        """
        dic = [set() for _ in range(n)]
        visited = [False for _ in range(n)]
        for x,y in edges:
            dic[x].add(y)
            dic[y].add(x)

        def dfs(start):
            for i in dic[start]:
                if not visited[i]:
                    visited[i] = True
                    dfs(i)
        result = 0
        for i in range(n):
            if not visited[i]:
                dfs(i)
                result += 1
        return result

File: Java/323__Number_of_Connected_Components_in_an_Undirected_Graph.java
class Solution {
    class DSU {
        int[] parent;

        DSU (int size) {
            parent = new int[size];
            for (int i = 0; i < size; i++)
                parent[i] = i;
        }

        public int find(int x){
            if (x != parent[x])
                parent[x] = find(parent[x]);
            return parent[x];
        }

        public void union(int x, int y){
            int xr = find(x), yr = find(y);
            parent[xr] = yr;
        }
    }

    public int countComponents(int n, int[][] edges) {
        DSU dsu = new DSU(n);
        for (int[] edge : edges)
            dsu.union(edge[0], edge[1]);
        int result = 0;
        for (int i = 0; i < n; i++)
            if (dsu.find(i) == i)
                result++;
        return result;
    }
}
